File: library/src/test/java/eu/codlab/testpromise/resolve/PromiseTestWithSumDelayed.java
package eu.codlab.testpromise.resolve;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import org.junit.Before;
import org.junit.Test;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

import eu.codlab.simplepromise.ErrorPromise;
import eu.codlab.simplepromise.Promise;
import eu.codlab.simplepromise.PromiseExec;
import eu.codlab.simplepromise.PromiseSolver;
import eu.codlab.simplepromise.Solver;
import eu.codlab.testpromise.utils.AndroidMockUtil;

/**
 * Created by kevinleperf on 06/04/2018.
 */

public class PromiseTestWithSumDelayed {

    @Before
    public void setHandler() {
        Promise.setHandler(AndroidMockUtil.mockMainThreadHandler());
    }

    @Test
    public void test() throws InterruptedException {
        final CountDownLatch latch = new CountDownLatch(1);
        final int[] final_result = {0};

        System.out.println("executing test");
        execute()
                .then(new PromiseExec<Integer, Integer>() {
                    @Override
                    public void onCall(@Nullable final Integer result, @NonNull final Solver<Integer> solver) {
                        new Thread() {
                            @Override
                            public void run() {
                                System.out.println("sleeping...");
                                try {
                                    sleep(2000);
                                } catch (InterruptedException e) {
                                    e.printStackTrace();
                                }
                                solver.resolve(10 + result);
                            }
                        }.run();
                    }
                })
                .then(new PromiseExec<Integer, Integer>() {
                    @Override
                    public void onCall(@Nullable Integer result, @NonNull Solver<Integer> solver) {
                        final_result[0] = 10 + result;
                        latch.countDown();
                    }
                })
                .error(new ErrorPromise() {
                    @Override
                    public void onError(@NonNull Throwable error) {
                        System.out.println("error catched");
                        final_result[0] = 0;
                        error.printStackTrace();
                        latch.countDown();
                    }
                });

        //6s are enough
        latch.await(6, TimeUnit.SECONDS);
        if (final_result[0] != 30) {
            throw new IllegalStateException("Expected 30... having " + final_result[0]);
        } else {
            System.out.println("having result " + final_result[0]);
        }
    }

    private Promise<Integer> execute() {
        return new Promise<>(new PromiseSolver<Integer>() {
            @Override
            public void onCall(@NonNull Solver<Integer> solver) {
                solver.resolve(10);
            }
        });
    }
}

File: promise/src/test/kotlin/com/voxeet/testpromise/subpromise/promise_io/PromiseTestWithSumDelayed.kt
package com.voxeet.testpromise.subpromise.promise_io

import com.voxeet.promise.HandlerFactory
import com.voxeet.promise.Promise
import com.voxeet.promise.solve.PromiseSolver
import com.voxeet.promise.solve.Solver
import com.voxeet.promise.solve.ThenPromise
import com.voxeet.testpromise.mockedhandler
import org.junit.Before
import org.junit.Test
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit

class PromiseTestWithSumDelayed {
    @Before
    fun setHandler() {
        HandlerFactory.setHandler(mockedhandler())
    }

    @Test
    @Throws(InterruptedException::class)
    fun test() {
        val latch = CountDownLatch(1)
        val final_result = intArrayOf(0)
        val resolve_later = Promise(PromiseSolver { solver: Solver<Int?> ->
            object : Thread() {
                override fun run() {
                    println("sleeping...")
                    try {
                        sleep(2000)
                    } catch (e: InterruptedException) {
                        e.printStackTrace()
                    }
                    solver.resolve(10)
                }
            }.run()
        }).then { result: Int?, solver: Solver<Int?> ->
            solver.resolve(
                result!! * 10
            )
        }
        println("executing test")
        execute()
            .then(ThenPromise<Int?, Int?> { result: Int? ->
                Promise { solver ->
                    solver.resolve(resolve_later
                        .then { new_result: Int?, solver1: Solver<Int?> ->
                            solver1.resolve(
                                new_result!! + result!!
                            )
                        }
                    )
                }
            })
            .then { final_result[0] = it!! }
            .error { error: Throwable ->
                println("error catched")
                final_result[0] = 0
                error.printStackTrace()
                latch.countDown()
            }

        //6s are enough
        latch.await(6, TimeUnit.SECONDS)
        check(final_result[0] == 110) { "Expected 110... having " + final_result[0] }
        println("having result " + final_result[0])
    }

    private fun execute(): Promise<Int?> {
        return Promise { solver: Solver<Int?> -> solver.resolve(10) }
    }
}
